attribute.go

```diff
@@ -107,26 +107,27 @@ func MarshalAttributes(attrs []Attribute) ([]byte, error) {
 // It is recommend to use the AttributeDecoder type where possible instead of calling
 // UnmarshalAttributes and using package nlenc functions directly.
 func UnmarshalAttributes(b []byte) ([]Attribute, error) {
-	var attrs []Attribute
-	var i int
-	for {
-		if i > len(b) || len(b[i:]) == 0 {
-			break
-		}
 
-		var a Attribute
-		if err := (&a).unmarshal(b[i:]); err != nil {
-			return nil, err
-		}
+	ad, err := NewAttributeDecoder(b)
+	if err != nil {
+		return nil, err
+	}
 
-		if a.Length == 0 {
-			i += nlaHeaderLen
-			continue
-		}
+	// Return a nil slice when there are no attributes to decode.
+	if ad.Len() == 0 {
+		return nil, nil
+	}
 
-		i += nlaAlign(int(a.Length))
+	attrs := make([]Attribute, 0, ad.Len())
+
+	for ad.Next() {
+		if ad.attr().Length != 0 {
```

Comment:
I seem to remember some weirdness around flag attributes where they might have length 0. I don't think we should filter these probably. /cc @florianl

Reply:
Let's follow up in an issue as discussed: https://github.com/mdlayher/netlink/blob/master/attribute_test.go#L285

As it stands, `UnmarshalAttributes` does not emit any zero-length attributes, so this clause was added to match existing behaviour.

Reply:
i've run tests for q_dsmark (where attribute type flag is used) with this PR and all looks good :+1: